### ISSUE TITLE
Consider allowing to set remote span as parent when instrumenting RabbitMQ

### ISSUE THREAD

**Issue by josh-romanowski** [rabbitmq/rabbitmq-dotnet-client#1666]
### Is your feature request related to a problem? Please describe.

In the past we had our own implementation where we implemented distributed tracing over RabbitMQ. So we were really excited to see native support of distributed tracing in the library.

We modeled our spans in a way where the span in the publisher is the parent of the spans created in consumers of a message. And we would like to continue modeling our messages in that way.

While the instrumentation library code [here](https://github.com/rabbitmq/rabbitmq-dotnet-client/blob/ee6646584f675238f0ee67c459dffd03c42abbc6/projects/RabbitMQ.Client.OpenTelemetry/TraceProviderBuilderExtensions.cs#L21) suggests that a parent child relationship is intended in the library, and the implementation [here](https://github.com/rabbitmq/rabbitmq-dotnet-client/blob/ee6646584f675238f0ee67c459dffd03c42abbc6/projects/RabbitMQ.Client/client/impl/RabbitMQActivitySource.cs#L148) also suggests that at least the possibility to set the extracted remote context as **parent**, the implementation currently only allows to set the extracted context as **link**. Also, the current interface for the `ContextExtractor` also does not allow to overwrite the `OpenTelemetryContextExtractor` in a way that would allow setting the extracted context as parent.

### Describe the solution you'd like

Instead of adding the remote span as a link to the span in a consumer, add the span as a parent.

### Describe alternatives you've considered

If it is not wanted to always set the remote span as a parent by default (see semantic conventions), at least give the possibility to overwrite the `ContentExtractor` in a way that allows setting either/or both parent and link based on the implementation. This would at least allow us to write our own extractor that sets the span as a parent.

### Additional context

[Semantic conventions for messaging systems](https://opentelemetry.io/docs/specs/semconv/messaging/messaging-spans/#context-propagation)

**Comment by lukebakken** [comment]
I will assume that any changes needed will not affect the public API for `RabbitMQ.Client`. I've put this in the 7.1.0 milestone for that reason.

**Comment by Tornhoof** [comment]
Yeah making it configurable would be nice, our own implementation also used a parent-child relation and at least for Grafana it's a break in the flow If you follow the traces. I'd then switch back to parent-child too.

**Comment by josh-romanowski** [comment]
As it appears the idea of allowing also to use the context as parent I can have look at making a MR for that.

@lukebakken I would assume making that configurable would at least require a small addition to the API of `RabbitMQActivitySource`. Would that be okay with you?

And the other question would be how you envision the configuration for that? Thinking about it probably adding a flag for what to use the extracted context for is more appropriate than actually extending `ContentExtractor` since I would assume that you would always use the same extracted context for both relations. What is you opinion?

**Comment by lailabougria** [comment]
I tend to agree with @Tornhoof, I would link by default, adhering to the conventions as @lmolkova pointed out, but making it configurable could help users who prefer having the parent-child relationship.

**Comment by lmolkova** [comment]
setting parent by default (in addition to link) is fine too - that's what we're going to do (or do already) in Azure ServiceBus EventHubs SDKs.

I believe @martinjt has the case for turning parent-child off, but (sorry Martin) he's the only one I heard from who wants links only. At the same time we've got overwhelming feedback to set parent and have one trace.

**Comment by lukebakken** [comment]
@josh-romanowski a PR would be welcome and yes, now is the time for API changes.

**Comment by martinjt** [comment]
The case for parent-child revolves around the idea that messages are sync, or close to sync. Which isn't always the case, and in my talking to a lot of Event Driven Architecture people, they confirmed that unless you're doing synchronous/transactional messaging, you shouldn't assume that.

With that in mind, having a parent/child relationship where the gap could be days break most trace waterfall views, and isn't really in keeping with the ideas behind how you model traces.

So if you're doing sync or close to synchronous messaging, then it does make sense to make them parent/child, but the default shouldn't be that, since it's then a different context you're working in, and you're not moving back up the chain with a response.

To be clear though, I've advocated for it to be configurable by the implementer. It shouldn't be both, you should be able to define, at a message type level, whether this invocation is part of a transaction/request/context, or whether this should start a new context and link, and ultimately, even whether it should be linked at all.

My personal view is that all telemetry is you, as the developer who knows your system, modelling your debug experience to make sense, therefore it should be optional, configuration, and with sensible defaults that will scale. Using links here is what should be happening by default, but the fact that it's not configurable is a super valid enhancement.

(I'm slightly worried that it's not going to be like that in the ServiceBus SDKs now @lmolkova, which would be a mistake)

**Comment by lailabougria** [comment]
@lmolkova @martinjt For context, we've recently made changes to NServiceBus.

The approach we have now is as follows:
- If you send a message, you get parent-child - the receive operation is a child of the send operation
- If you publish a message, you get links - a publish is a fan-out and has an unknown amount of subscribers, which is why we link.
- If you send a delayed message (delayed delivery), you get links - this is guaranteed to happen with a delay, so a link makes sense to limit the duration of the trace

Apart from the last option, where we enforce links, we've provided APIs that allow you to deviate per send/publish to alter the behavior. So if you prefer a parent-child when publishing, you can configure that. And when you prefer linking for a send, you can do that too. 

We've also seen most users ask for parent-child relationships. However, we've also gotten feedback that when everything is parent-child everywhere, you get a never-ending trace that's way too hard to understand. In a message-based system, almost every action can lead to a follow-up send/publish operation, making traces very long. This is why we provided APIs that allow users to deviate at the operation level, as they have the best context of what makes sense to include/exclude from the trace.

**Comment by SzymonPobiega** [comment]
Let me add my two cents also

My understanding is that RabbitMQ is a bit of a special type of a messaging broker because it has some unique features that make it well suited for synchronous request-reply communication. The tricky bit is, the knowledge if a given message is supposed to be considered async or sync is available only at the site when it is created, likely at the level of a higher-level framework, such as MassTransit or NServiceBus.

Expanding on what @lailabougria wrote, the parent/link decision process in NServiceBus happens entirely on the sending side, following the logic I outlined above: it is the sender who has the context to decide, for any given message. In addition to that, NServiceBus will prioritize the decision made by the messaging SDK over its own link/parent logic _if the SDK does support OpenTelemetry and if NServiceBus transport adapter has been updated to take advantage of that_.

Based on that, if messaging SDK does not support OpenTelemetry, the decision to link/parent is made on the sending side with defaults being link for events and parent for commands. 

If, however, the SDK does support OpenTelemetry, NServiceBus will ignore the hints of the sending framework side and use whatever strategy SDK receiver employed. 

Now, coming back to my first point about the sync/async context being available only on the sending side, I would like to propose a following design:

- The sending code (either the code that calls RabbitMQ directly or a code of a higher-level framework such, as MassTransit or NServiceBus) decides if a given message is to be treated as sync or async with respect to tracing. Defaults. such as events are async while commands are sync. apply here. The decision is somehow encoded in the message that is passed to the RabbitMQ SDK
- RabbitMQ SDK sending side encodes this decision in the message that is dispatched. General overrides such as "always use parent" or "never use parent" might apply here. 
- RabbitMQ SDK receiving side decodes the link/parent decision and creates activity accordingly
- The receiving code (either the code that calls RabbitMQ directly or a code of a higher-level framework, such as MassTransit or NServiceBus) uses that activity as a parent

This would allow the higher-level frameworks to make sensible decisions for the users on the receiving side regardless if the broker SDK has the OpenTelemetry support enabled or disabled.

**Comment by josh-romanowski** [comment]
I really appreciate the detailed discussion.

However, I have to admit that some of the points are beyond my knowledge of how to implement them in code here. A general overwrite would already meet the requirements I would have for a solution.

If that's ok with everyone, I'll provide a simple PR that allows a general override. For a more detailed configuration, I'd ask you guys to expand on the suggestion then. I'll be on FTO starting tomorrow and unfortunately won't have the capacity to deal with the details of a more detailed configuration until then.

**Comment by ngbrown** [comment]
I just ran into this issue. I have a system that does both fan-out and RPC, so with the RPC calls, I want the `Activty.Parent` to be populated from the `traceparent` header value. Currently, the result of `ContextExtractor()` is being passed into `StartLinkedRabbitMQActivity(linkContext: )` and `parentContext` is never passed from any of the call sites.

My traces currently look like this and there are three separate top level trace id's when I just want one (currently the three are: send request, handle request, handle response).

![Screenshot 2024-11-08 205126](https://github.com/user-attachments/assets/7ee0055f-9662-4c22-92a6-894cb8a1fc78)

Without taking into account the fan-out case where activity links might be preferred, I have a branch where I just move the context to `parentContext` and get the traces that I want.

branch: https://github.com/ngbrown/rabbitmq-dotnet-client/blob/fix-activity-parent commit: https://github.com/ngbrown/rabbitmq-dotnet-client/commit/5b713870054dfb939a75902b7202b27de1d2152c

![Screenshot 2024-11-08 204741](https://github.com/user-attachments/assets/4816ba66-ff40-48be-9d2e-4ec19ee331da)

So this should be configurable in some way. Probably on a per-message basis, and if there is an automated way to tell if the message came from a topic exchange that would be ideal (since that's the only way I pass fan-out messages). For real-time fan-out messages, even to lots of receivers, I would still rather have the normal parent/child relationship.

As an aside, this is using Serilog/SerilogTracing with the Seq web service. The https://github.com/serilog-tracing/serilog-tracing/pull/153 is still pending and my RabbitMQ adaptor for SerilogTracing is here: https://github.com/ngbrown/grpc-testing/tree/with-serilog-tracing-pr/SerilogTracing.Instrumentation.RabbitMQ.

Even if SerilogTracing understood Activity linking, the linked trace id keeps changing, so indexing on that isn't really that useful.

**Comment by eerhardt** [comment]
Just adding a +1 here to being able to have a parent-child relationship for the "Receive/Process" span. With the [.NET Aspire Dashboard](https://learn.microsoft.com/dotnet/aspire/fundamentals/dashboard/overview), having the parent-child relationship makes the Tracing much easier to consume/understand.

![image](https://github.com/user-attachments/assets/4a782615-4c32-4284-b0af-cdc1fc9062e1)

The "Links" show up in the bottom of the details of a span, but they are hard to find and see in one shot:

![image](https://github.com/user-attachments/assets/7d5884f9-aaad-4141-8e96-daf2fbece977)

![image](https://github.com/user-attachments/assets/f1c6a3ed-746c-4125-8435-4f063a2c693d)

Compare this with the experience with Azure ServiceBus, where the order processing gets nested in the whole trace that kicked it off:

![image](https://github.com/user-attachments/assets/833cc01d-303d-49d7-acb5-11a9e8c80047)

ServiceBus also has Links, so you can see them as well:

![image](https://github.com/user-attachments/assets/2e24e5ae-6012-4221-ae53-2352df58098b)

My 2 cents (for what it is worth) is I think for a normal `BasicPublishAsync` => `AsyncEventingBasicConsumer` flow, having the parent-child relationship plus Links by default makes sense, and then having an option to disable the parent-child relationship, if necessary.

**Comment by lukebakken** [comment]
@josh-romanowski @eerhardt @ngbrown - any chance of a PR "soon"? I'd like to release 7.1.0 by the end of Dec 2024.

**Comment by meum** [comment]
If no one else is working on this, I could make a simple PR.
Is it enough with a boolean to switch between using `StartRabbitMQActivity` and `StartLinkedRabbitMQActivity`?

**Comment by stebet** [comment]
My two cents here.

Considering that links are the default for consumer spans in the [OTel Messaging conventions](https://opentelemetry.io/docs/specs/semconv/messaging/messaging-spans/#consumer-spans), I'd say keep link as the default with an option to opt-in to them being parent-child instead.

> A single “Process” or “Receive” span can account for a single message, for a batch of messages, or for no message at all (if it is signalled that no messages were received). For each message it accounts for, the “Process” or “Receive” span SHOULD link to the message’s creation context.

There are also considerations [in the spec](https://opentelemetry.io/docs/specs/semconv/messaging/messaging-spans/#message-creation-context-as-parent-of-process-span) when the messaging library *knows* that there will be a parent-child relationship but I don't see how that can be easily achieved in the library unless there is some header or option in the message publishing call that can signal that in some way. I don't think the API should make assumptions other than what the spec recommends in general.

> Exclusively for single messages scenarios, the “Process” span MAY use the message’s creation context as its parent, thus achieving a direct parent-child relationship between producer and consumer(s). Instrumentations SHOULD document whether they use the message creation context as a parent for “Process” spans and MAY provide configuration options allowing users to control this behavior.
> 
> It is NOT RECOMMENDED to use the message creation context as the parent of “Process” spans (by default) if processing happens in the scope of another span.
> 
> If instrumentation uses the message creation context as the parent for “Process” spans in the scope of another valid ambient context, it SHOULD add the ambient context as a link on the “Process” span to preserve the correlation between message processing and that context.

**Comment by meum** [comment]
I made a PR for this, any feedback is welcome.
